common/Uri.cpp

```diff
@@ -0,0 +1,69 @@
+/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4; fill-column: 100 -*- */
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <config.h>
+
+#include "Uri.hpp"
+#include <common/Util.hpp>
+#include <common/Log.hpp>
+
+#include <Poco/URI.h>
+
+const std::string Uri::Reserved = ",/?:@&=+$#";
+
+std::string Uri::encode(const std::string& uri, const std::string& reserved)
+{
+    std::string encoded;
+    Poco::URI::encode(uri, reserved, encoded);
+    return encoded;
+}
+
+std::string Uri::decode(const std::string& uri)
+{
+    std::string decoded;
+    Poco::URI::decode(uri, decoded);
+    return decoded;
+}
+
+bool Uri::needsEncoding(const std::string& uri, const std::string& reserved)
+{
+    const std::string decoded = decode(uri);
+    if (decoded != uri)
+    {
+        // We could decode it; must have been encoded already.
+        return false;
+    }
+
+    // Identical when decoded, might need encoding.
+    const std::string encoded = encode(uri, reserved);
+
+    // If identical, then doesn't need encoding.
+    return encoded != uri;
+}
+
+std::string Uri::getFilenameFromURL(const std::string& url)
+{
+    const auto [base, filename, ext, params] = Util::splitUrl(url);
```

Comment:
The proper (modern) way is by marking the variable in question with `[[maybe_unused]]`. But in this case, we wouldn't want to mark _all_ these variables as possibly-unused, which is I think the only option here.

That is, `const auto [ [[maybe_unused]] base, filename, ext, params] = Util::splitUrl(url);` is invalid.
But we don't want the alternative, which is valid: `[[maybe_unused]] const auto [ base, filename, ext, params] = Util::splitUrl(url);`

So that leaves classic void-casting as an option. I don't know if there are other solutions, though.